include/drivers/gpio/gpio_esp32.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __GPIO_ESP32_H__
+#define __GPIO_ESP32_H__
+
+struct enable_bits {
+	int reg;
+	int mask;
+};
+
+struct peripheral {
+	struct enable_bits clk;
+	struct enable_bits rst;
+};
+
+static inline void gpio_esp32_set_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) | v, mem_addr);
+}
+
+static inline void gpio_esp32_clear_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
+}
+
+static void gpio_esp32_enable_peripheral(const struct peripheral *peripheral)
```

Comment:
This function could be moved to `arch/xtensa/soc/esp32/peripheral.h` or something like this.